File: leetcode/problems/language-specific/go/230.kth-smallest-element-in-a-bst.go
/*
 * @lc app=leetcode id=230 lang=golang
 *
 * [230] Kth Smallest Element in a BST
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func kthSmallest(root *TreeNode, k int) int {
	curIndex := 0
	kthNode := inOrder(root, k, &curIndex)
	return kthNode.Val
}

func inOrder(root *TreeNode, k int, curIndex *int) *TreeNode {
	if root == nil {
		return root
	}

	left := inOrder(root.Left, k, curIndex)
	if left != nil {
		return left
	}

	*curIndex++
	if *curIndex == k {
		return root
	}

	return inOrder(root.Right, k, curIndex)
}

// @lc code=end

File: leetcode/problems/language-specific/python/230.kth-smallest-element-in-a-bst.py
#
# @lc app=leetcode id=230 lang=python3
#
# [230] Kth Smallest Element in a BST
#

# @lc code=start

# * Iterative Inorder Traversal Solution | O(n) Time | O(h) Space
# * n -> The number of nodes in the tree | h -> The height of the tree

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        stack = []
        cur = root
        while cur or stack:
            # * Move left until we've a valid node.
            while cur:
                stack.append(cur)
                cur = cur.left

            # * Pop the topmost element from the stack.
            cur = stack.pop()
            k -= 1
            if not k:
                return cur.val

            # * Move right.
            cur = cur.right

# @lc code=end
